### ISSUE TITLE
S3242: False positive 

### ISSUE THREAD

**Issue by antoinebj** [SonarSource/sonar-dotnet#2479]
### Description

A false positive exists for rule S3242 when using an indexer in a deconstructing expression.

### Repro steps

Run analysis of the following code:

```csharp
public sealed class S3242
{
    public void SomeMethod(IReadOnlyList<S3242DeconstructibleType> list) // S3242 suggesting IReadOnlyCollection
    {
        for (var i = 0; i < list.Count; ++i)
        {
            var (key, value) = list[i];
        }
    }
}

public class S3242DeconstructibleType
{
    public void Deconstruct(out object key, out object value)
    {
        key = value = null;
    }
}
```

Notice that the indexer of the `list` variable is used, but the message for the rule violation suggests changing its type to `IReadOnlyCollection`, which does not have an indexer and would create a compilation error.

### Expected behavior

No violation of rule S3242 reported.

### Actual behavior

A violation of rule S3242 is reported.

### Known workarounds

Suppression of the message.

### Related information

* SonarC# Version 7.15.0.8572
* Visual Studio Version 16.1.4

**Comment by andrei-epure-sonarsource** [comment]
thanks @antoinebj  I confirm the FP

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1988](https://sonarsource.atlassian.net/browse/NET-1988)

[NET-1988]: https://sonarsource.atlassian.net/browse/NET-1988?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
